src/librustc_mir/transform/promote_consts.rs

```diff
@@ -132,22 +155,603 @@ impl<'tcx> Visitor<'tcx> for TempCollector<'tcx> {
+                            let ty = Place::ty_from(place.base, &proj.base, self.body, self.tcx)
+                                .projection_ty(self.tcx, &proj.elem)
+                                .ty;
+                            if ty.is_freeze(self.tcx, self.param_env, DUMMY_SP) {
+                                has_mut_interior = false;
+                            }
+                        }
+
+                        place_projection = &proj.base;
+                    }
+                }
+                if has_mut_interior {
+                    return Err(Unpromotable);
+                }
+
+                self.validate_place(place)
+            }
+
+            Rvalue::Aggregate(_, ref operands) => {
+                for o in operands {
+                    self.validate_operand(o)?;
+                }
+
+                Ok(())
+            }
+        }
+    }
+
+    fn validate_call(
+        &self,
+        callee: &Operand<'tcx>,
+        args: &[Operand<'tcx>],
+    ) -> Result<(), Unpromotable> {
+        if !self.explicit && self.is_non_const_fn {
+            if let ty::FnDef(def_id, _) = callee.ty(self.body, self.tcx).sty {
+                // Never promote runtime `const fn` calls of
+                // functions without `#[rustc_promotable]`.
+                if !self.tcx.is_promotable_const_fn(def_id) {
+                    return Err(Unpromotable);
+                }
+            }
+        }
+
+        let fn_ty = callee.ty(self.body, self.tcx);
+        match fn_ty.sty {
+            ty::FnDef(def_id, _) => {
+                match self.tcx.fn_sig(def_id).abi() {
+                    Abi::RustIntrinsic |
+                    Abi::PlatformIntrinsic => {
+                        assert!(!self.tcx.is_const_fn(def_id));
+                        match &self.tcx.item_name(def_id).as_str()[..] {
+                            | "size_of"
+                            | "min_align_of"
+                            | "needs_drop"
+                            | "type_id"
+                            | "bswap"
+                            | "bitreverse"
+                            | "ctpop"
+                            | "cttz"
+                            | "cttz_nonzero"
+                            | "ctlz"
+                            | "ctlz_nonzero"
+                            | "wrapping_add"
+                            | "wrapping_sub"
+                            | "wrapping_mul"
+                            | "unchecked_shl"
+                            | "unchecked_shr"
+                            | "rotate_left"
+                            | "rotate_right"
+                            | "add_with_overflow"
+                            | "sub_with_overflow"
+                            | "mul_with_overflow"
+                            | "saturating_add"
+                            | "saturating_sub"
+                            | "transmute"
+                            => return Err(Unpromotable),
```

Comment:
Is this a *blacklist* of unpromotable intrinsics, or what is going on here?

Cc https://github.com/rust-lang/rust/issues/61495

Reply:
It's copy-pasted from `qualify_consts`.

In general I can attempt to do more deduplication, and for that it would probably be best to make a `consts` module that handles qualification/checking/promotion, so that I can define a lot of `pub(super)` things.

But I wanted to have minimal reuse for the crater run to avoid "unaccounted for" logic.

Reply:
Sure. I was just surprised by the polarity here. I expected a whitelist, but this looks like a blacklist.

Reply:
In fact I think this list should just go. We don't want to promote any intrinsics.

The only intrinsic that stable code can directly access is `transmute`, and we certainly don't want to promote that. All of the others are exposed via wrapping functions, and we use `#[rustc_promotable]` to control their promotability.

Reply:
Yes, this match is unnecessary. It can just mark any intrinsics as unpromotable